### ISSUE TITLE
Conversion script is no longer working

### ISSUE THREAD

**Issue by Atcold** [e-lab/imagenet-multiGPU.torch#1]
The `convert-model.lua` script is no longer working with the new `DataParallelTable()` construct, which exist only in `cunn`.

Error stack

``` lua
/usr/local/bin/luajit: /usr/local/share/lua/5.1/cunn/DataParallelTable.lua:485: type() not supported for DataParallelTable.
stack traceback:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/cunn/DataParallelTable.lua:485: in function 'type'
        /usr/local/share/lua/5.1/nn/utils.lua:45: in function 'recursiveType'
        /usr/local/share/lua/5.1/nn/utils.lua:41: in function 'recursiveType'
        /usr/local/share/lua/5.1/nn/Module.lua:123: in function 'float'
        convert-model.lua:265: in function 'convert_model'
        convert-model.lua:293: in main chunk
        [C]: in function 'dofile'
        /usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00406260
```

P.S. A network can be found here

``` bash
GPU6:/media/HDD1/model-cache/results-5k10k/20151113-163408-alexnetowtbn\,epochNumber=52\,nDonkeys=8\,nEpochs=100000000\,nGPU=4\,normalize=f/model_68.t7
```

**Comment by jhjin** [comment]
Fixed via https://github.com/e-lab/imagenet-multiGPU.torch/commit/f96d3ebc92ddc8d0430dd722bad53a01ca378df5. I suggest you to check if it gives the same numerical result as the one from the original. Merge to the master if it works as expected.

**Comment by Atcold** [comment]
Hmm...

``` lua
net = torch.load('model.t7')
inp = torch.FloatTensor(8, 3, 224, 224)
for i = 1, 8 do inp[i] = image.scale(image.lena(), 224, 224):float() end
out = net:forward(input)
print(out[1][{ {1, 10} }]:view(1, -1))
0.0001 *                                                                        
  0.0665  0.0428  0.0987  0.0698  4.4356  0.0137  7.8518  0.0849  0.2295  0.0476
[torch.FloatTensor of size 1x10]
```

and

``` lua
model = torch.load('model_68.t7')
for i, module in ipairs(model.modules[1].modules) do cutorch.setDevice(model.modules[1].gpuAssignments[i]); model.modules[1].modules[i] = module:float():cuda(); end
cutorch.setDevice(1)                                                                                                                                                                                                                                                        
inp = torch.CudaTensor(8,3,224,224)
for i = 1, 8 do
   inp[i] = image.scale(image.lena(), 224, 224):cuda()
end
out = model:forward(inp)
print(out[1][{ {1, 10} }]:view(1, -1))
Columns 1 to 8
 -9.9937 -10.8176  -9.1500  -8.3259  -9.4792 -11.1297  -7.3883 -10.0991

Columns 9 to 10
 -8.4595 -10.4629
[torch.CudaTensor of size 1x10]
```

**Comment by jhjin** [comment]
@Atcold Did you replace the last layer of `model.t7` with `nn.LogSoftMax`?

**Comment by Atcold** [comment]
Did nothing. Converted and tested.
LOL, just a sec...

``` lua
th> out[1][{ {1, 10} }]:view(1, -1):log()                              
Columns 1 to 8                                                         
-11.9214 -12.3611 -11.5263 -11.8729  -7.7207 -13.5018  -7.1496 -11.6763

Columns 9 to 10                                                        
-10.6821 -12.2560                                                      
[torch.FloatTensor of size 1x10]

th> out[1][{ {1, 10} }]:view(1, -1)
Columns 1 to 8
 -9.9937 -10.8176  -9.1500  -8.3259  -9.4792 -11.1297  -7.3883 -10.0991

Columns 9 to 10
 -8.4595 -10.4629
[torch.CudaTensor of size 1x10]
```

**Comment by jhjin** [comment]
Please do that. Your numbers are all negative.

**Comment by Atcold** [comment]
I've edited the answer. Plz, check the new update.

**Comment by jhjin** [comment]
Could you slice the output before `nn.SoftMax` and compare it again? softmax/logsoftmax is sensitive to precision and the gap might be caused by `cunn`/ `nn`'s different behavior (since you have many classes). It looks like correct because the order is preserved. Otherwise, I will look at it later.

**Comment by Atcold** [comment]
What do you think, is it correct?

``` lua
th> net.modules[19]
nn.Linear(4096 -> 5000)
                                                                      [0.0003s]
th> net.modules[19].output[1][{ {1,10} }]:view(1, -1)
-0.9888 -1.4375 -0.5868 -0.9395  3.2480 -2.6085  3.8200 -0.7393  0.2691 -1.3301
[torch.FloatTensor of size 1x10]

                                                                      [0.0005s]
th> model.modules[2].modules[4]
nn.Linear(4096 -> 5000)
                                                                      [0.0001s]
th> model.modules[2].modules[4].output[1][{ {1,10} }]:view(1, -1)
-0.9832 -1.8071 -0.1395  0.6846 -0.4687 -2.1192  1.6222 -1.0886  0.5510 -1.4524
[torch.CudaTensor of size 1x10]

                                                                      [0.0004s]
```

I'm going to make a script. Perhaps I'm comparing different models...

**Comment by Atcold** [comment]
OK, [here](https://github.com/e-lab/imagenet-multiGPU.torch/blob/DataParallelTable-compat/tools/conversionTest.lua) is the network comparison script.
You probably want to change `floatModelPath` to same location that you have writing permissions for.

**Comment by Atcold** [comment]
@jhjin, I have successfully removed the `DataParallelTable()` construct. The errors in the output come from a bad conversion `cudnn` -> `float`.

**Comment by jhjin** [comment]
As agreed, conversion is fine, the underlying problem was the error in the training script, especially in retraining. Fixed by https://github.com/e-lab/imagenet-multiGPU.torch/commit/07bd14abf7c1281199275f439e29767f19915df6.
